File: sdk/ts/src/price.ts
/**
 * Human-facing `Price` codec.
 *
 * The on-chain `Price` is a `u32` decimal floating-point comparison key
 * (see programs/dropset/src/price.rs and architecture.md § Price). The
 * IDL exposes it only as raw `u32` bits — the generated `Price` type is
 * `number` (those bits) — so this module layers the bits <-> decimal
 * conversion the frontend needs to display prices and to build
 * `set_reference_price` / `swap` arguments.
 *
 * Layout: `[5-bit biased exponent][27-bit normalized significand]`.
 * The represented value is `significand * 10^(unbiasedExponent - 7)`,
 * with the significand normalized to 8 significant digits
 * (`10_000_000..=99_999_999`) and the exponent biased by 16.
 */

const SIGNIFICAND_BITS = 27;
const SIGNIFICAND_MASK = (1 << SIGNIFICAND_BITS) - 1;
const BIAS = 16;
const SIGNIFICAND_MIN = 10_000_000;
const SIGNIFICAND_MAX = 99_999_999;
const MAX_BIASED_EXPONENT = 31;

/** Market-sell sentinel — no minimum fill price. Raw bits `0x00000000`. */
export const PRICE_ZERO = 0;
/** Market-buy sentinel — no maximum fill price. Raw bits `0xFFFFFFFF`. */
export const PRICE_INFINITY = 0xffff_ffff;

/** Raw `Price` bits, as carried in instruction args, accounts, and events. */
export type PriceBits = number;

export function isZeroPrice(bits: PriceBits): boolean {
  return bits >>> 0 === PRICE_ZERO;
}

export function isInfinityPrice(bits: PriceBits): boolean {
  return bits >>> 0 === PRICE_INFINITY;
}

/** The lower 27 bits. Meaningful only for regular (non-sentinel) prices. */
export function priceSignificand(bits: PriceBits): number {
  return (bits >>> 0) & SIGNIFICAND_MASK;
}

/** The upper 5 bits. Meaningful only for regular prices. */
export function priceBiasedExponent(bits: PriceBits): number {
  return (bits >>> 0) >>> SIGNIFICAND_BITS;
}

/** `biasedExponent - 16`. Meaningful only for regular prices. */
export function priceUnbiasedExponent(bits: PriceBits): number {
  return priceBiasedExponent(bits) - BIAS;
}

/**
 * Whether `bits` is a valid encoding: a sentinel, or a regular price
 * whose significand sits in `[10_000_000, 99_999_999]`. Mirrors
 * `Price::is_valid` — invalid patterns mis-sort in the order book, so
 * callers must check user-supplied bits before submitting them.
 */
export function isValidPrice(bits: PriceBits): boolean {
  const u = bits >>> 0;
  if (u === PRICE_ZERO || u === PRICE_INFINITY) return true;
  const sig = priceSignificand(u);
  return sig >= SIGNIFICAND_MIN && sig <= SIGNIFICAND_MAX;
}

/** Assemble raw bits from a validated significand and **biased** exponent. */
export function priceFromParts(
  significand: number,
  biasedExponent: number,
): PriceBits {
  if (
    significand < SIGNIFICAND_MIN ||
    significand > SIGNIFICAND_MAX ||
    biasedExponent < 0 ||
    biasedExponent > MAX_BIASED_EXPONENT
  ) {
    throw new RangeError(
      `Price out of range: significand=${significand}, biasedExponent=${biasedExponent}`,
    );
  }
  return ((biasedExponent << SIGNIFICAND_BITS) | significand) >>> 0;
}

const SIGNIFICAND_MIN_BIG = BigInt(SIGNIFICAND_MIN);
const SIGNIFICAND_MAX_BIG = BigInt(SIGNIFICAND_MAX);

/**
 * Normalize a scaled integer significand + biased exponent into canonical
 * bits, mirroring `Price::from_scaled`. The significand is a `bigint` so
 * the normalization is exact integer math — matching Rust's `u64`
 * division bit-for-bit even past JS's 2^53 safe-integer range (callers
 * lift values that can exceed it; see {@link weightedAverage}). Digits
 * beyond the 8th are truncated toward zero. Throws if the result falls
 * outside the representable exponent range.
 */
function fromScaled(sig: bigint, biasedExp: number): PriceBits {
  if (sig === 0n) return PRICE_ZERO;
  let s = sig;
  let e = biasedExp;
  while (s > SIGNIFICAND_MAX_BIG) {
    s /= 10n;
    e += 1;
  }
  while (s < SIGNIFICAND_MIN_BIG) {
    s *= 10n;
    e -= 1;
  }
  if (e < 0 || e > MAX_BIASED_EXPONENT) {
    throw new RangeError(`Price exponent out of range: biasedExponent=${e}`);
  }
  // `s` is now in `[10_000_000, 99_999_999]`, well within safe-integer range.
  return priceFromParts(Number(s), e);
}

/**
 * Encode a decimal price (e.g. `1.085`) into raw `Price` bits.
 *
 * Intended for the FX value range the protocol targets (roughly
 * `1e-6 .. 1e6`); values are truncated to 8 significant digits. Throws a
 * `RangeError` on non-finite/negative input, and on a value so large that
 * `value * 1e7` would not fit in a `u64` — the boundary at which the Rust
 * `from_value` cast would otherwise silently saturate, so both forks
 * reject the same inputs. For the no-bound sentinels pass
 * {@link PRICE_ZERO} / {@link PRICE_INFINITY} directly.
 */
export function encodePrice(value: number): PriceBits {
  if (!Number.isFinite(value) || value < 0) {
    throw new RangeError(`Price must be a finite, non-negative number: ${value}`);
  }
  if (value === 0) return PRICE_ZERO;
  const scaled = value * 1e7;
  // Mirror Rust `from_value`: reject what would overflow the `u64` scaling
  // intermediate. `2 ** 64` is `u64::MAX as f64` (which rounds up to 2^64),
  // so any `scaled` below it truncates losslessly toward zero into range.
  if (scaled >= 2 ** 64) {
    throw new RangeError(`Price out of range: ${value}`);
  }
  // value = significand * 10^(unbiased - 7); choose unbiased = 0
  // (biased = BIAS) so the scaled integer is value * 1e7, then let
  // fromScaled renormalize the significand into the canonical range.
  return fromScaled(BigInt(Math.trunc(scaled)), BIAS);
}

/** Decode raw `Price` bits to a JS number. Sentinels map to `0` / `Infinity`. */
export function decodePrice(bits: PriceBits): number {
  const u = bits >>> 0;
  if (u === PRICE_ZERO) return 0;
  if (u === PRICE_INFINITY) return Number.POSITIVE_INFINITY;
  return priceSignificand(u) * 10 ** (priceUnbiasedExponent(u) - 7);
}

/** `quote_for_base(INFINITY)` / `base_for_quote(ZERO)` sentinel value. */
const U128_MAX = (1n << 128n) - 1n;
const U64_MAX = (1n << 64n) - 1n;

/** Clamp to `u128::MAX`, mirroring Rust's `saturating_*` on `u128`. */
const satU128 = (x: bigint): bigint => (x > U128_MAX ? U128_MAX : x);

/**
 * `base * price`, rounded toward zero — the exact integer math the on-chain
 * matcher uses (mirrors Rust `Price::quote_for_base`). `ZERO -> 0n`,
 * `INFINITY -> U128_MAX`. Use this (not {@link decodePrice} float math) for
 * cross-language-consistent ratios. `bigint` carries the full range; values
 * are not saturated to `u64` (lossless for the FX atom scales).
 */
export function quoteForBase(bits: PriceBits, base: bigint): bigint {
  const u = bits >>> 0;
  if (u === PRICE_ZERO) return 0n;
  if (u === PRICE_INFINITY) return U128_MAX;
  const sig = BigInt(priceSignificand(u));
  const unb = priceUnbiasedExponent(u) - 7;
  let x = base * sig;
  if (unb >= 0) for (let i = 0; i < unb; i++) x *= 10n;
  else for (let i = 0; i < -unb; i++) x /= 10n;
  return x;
}

/**
 * `quote / price`, rounded toward zero (mirrors Rust `Price::base_for_quote`).
 * `ZERO -> U128_MAX`, `INFINITY -> 0n`.
 */
export function baseForQuote(bits: PriceBits, quote: bigint): bigint {
  const u = bits >>> 0;
  if (u === PRICE_ZERO) return U128_MAX;
  if (u === PRICE_INFINITY) return 0n;
  const sig = BigInt(priceSignificand(u));
  const unb = priceUnbiasedExponent(u) - 7;
  let num = quote;
  let den = sig;
  if (unb >= 0) for (let i = 0; i < unb; i++) den *= 10n;
  else for (let i = 0; i < -unb; i++) num *= 10n;
  return den === 0n ? 0n : num / den;
}

/**
 * Shares-weighted average of two prices in *decoded-value* space (mirrors
 * Rust `Price::weighted_average`). Both prices are lifted to `value × 10^9`,
 * averaged by the integer weights, then re-encoded via {@link fromScaled}
 * (truncating toward zero — one ULP per merge, acceptable for cost-basis
 * bookkeeping). Backs the depositor cost-basis merge on a top-off deposit
 * ({@link mergeEntryBasis}).
 *
 * Returns `self` when both weights are zero; {@link PRICE_ZERO} if either
 * input is the ZERO sentinel; {@link PRICE_INFINITY} if either is INFINITY
 * — the engine's conservative sentinel collapse. Falls back to
 * {@link PRICE_ZERO} if the re-encode overflows the exponent range (Rust's
 * `unwrap_or(ZERO)`).
 */
export function weightedAverage(
  self: PriceBits,
  other: PriceBits,
  wSelf: bigint,
  wOther: bigint,
): PriceBits {
  if (wSelf === 0n && wOther === 0n) return self >>> 0;
  if (wSelf === 0n) return other >>> 0;
  if (wOther === 0n) return self >>> 0;
  if (isZeroPrice(self) || isZeroPrice(other)) return PRICE_ZERO;
  if (isInfinityPrice(self) || isInfinityPrice(other)) return PRICE_INFINITY;
  const SCALE = 1_000_000_000n;
  const vSelf = quoteForBase(self, SCALE);
  const vOther = quoteForBase(other, SCALE);
  // Mirror Rust's `u128` `saturating_mul` / `saturating_add`: clamp each
  // weighted product, their sum, and the weight total at `u128::MAX`. For
  // FX-range prices nothing saturates and this is a no-op, but for
  // structurally-valid prices above that band the products can exceed
  // `u128`, where plain bigint would otherwise diverge from the engine.
  const total = satU128(wSelf + wOther);
  const avg = satU128(satU128(wSelf * vSelf) + satU128(wOther * vOther)) / total;
  // value × 10^9 = sig × 10^(biased − 14) ⟹ the scaled input's biased
  // exponent is 14. Drive `fromScaled` from the `bigint` directly: `avg`
  // can exceed 2^53 (price value > ~9e6), where a `Number(avg)` round
  // would flip the 8th significand digit vs Rust's exact `u64` `from_scaled`.
  const avgU64 = avg > U64_MAX ? U64_MAX : avg;
  try {
    return fromScaled(avgU64, 14);
  } catch {
    return PRICE_ZERO;
  }
}

File: sdk/price-core/src/price.rs
//! `Price` codec — a solana-free port of the program's
//! `programs/dropset/src/price.rs`.
//!
//! `Price` is a `u32` decimal floating-point comparison key with 8
//! significant digits and a base-10 exponent, designed so raw unsigned
//! integer comparison matches price-value ordering. The IDL exposes it
//! only as raw `u32` bits (the on-chain type does not derive `IdlType`),
//! so the matching simulator and any off-chain renderer need this
//! decode/encode + the `quote_for_base` / `base_for_quote` ratio math.
//!
//! This is a hand-mirror of the on-chain logic. Per interface.md § SDK
//! (spine B), the long-term home for this math is a shared `price-core`
//! crate compiled to WASM with conformance vectors run in both Rust and
//! TS CI; until that lands, keep this in lockstep with the program copy.

const SIGNIFICAND_BITS: u32 = 27;
const SIGNIFICAND_MASK: u32 = (1 << SIGNIFICAND_BITS) - 1;
const BIAS: u8 = 16;
const SIGNIFICAND_MIN: u32 = 10_000_000;
const SIGNIFICAND_MAX: u32 = 99_999_999;
const MAX_BIASED_EXPONENT: u8 = 31;

/// A `u32` decimal floating-point comparison key. See the module docs and
/// architecture.md § Price.
#[repr(transparent)]
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct Price(u32);

impl core::fmt::Debug for Price {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match () {
            _ if self.is_zero() => write!(f, "Price::ZERO"),
            _ if self.is_infinity() => write!(f, "Price::INFINITY"),
            _ => write!(
                f,
                "Price(0x{:08X}, sig={}, exp={})",
                self.0,
                self.significand(),
                self.unbiased_exponent(),
            ),
        }
    }
}

impl Price {
    /// Market-sell sentinel — no minimum fill price.
    pub const ZERO: Self = Self(0);
    /// Market-buy sentinel — no maximum fill price.
    pub const INFINITY: Self = Self(u32::MAX);

    /// Wrap raw `u32` bits without validation. Call [`is_valid`](Self::is_valid)
    /// before use when the source is untrusted.
    #[inline(always)]
    pub const fn from_bits(bits: u32) -> Self {
        Self(bits)
    }

    /// The raw `u32` bit pattern.
    #[inline(always)]
    pub const fn as_u32(self) -> u32 {
        self.0
    }

    /// Encode a decimal value (e.g. `1.085`) into a `Price`. Intended for
    /// the FX value range (`value * 1e7` within f64 integer precision);
    /// truncates to 8 significant digits. `0.0` maps to [`Price::ZERO`].
    /// Returns `None` for non-finite/negative input or out-of-range
    /// exponent. Mirrors the TS `encodePrice`.
    pub fn from_value(value: f64) -> Option<Self> {
        if !value.is_finite() || value < 0.0 {
            return None;
        }
        if value == 0.0 {
            return Some(Self::ZERO);
        }
        Self::from_scaled((value * 1e7) as u64, BIAS as i16)
    }

    /// Construct from a validated significand and **biased** exponent.
    #[inline]
    pub const fn new(significand: u32, biased_exponent: u8) -> Option<Self> {
        if significand < SIGNIFICAND_MIN
            || significand > SIGNIFICAND_MAX
            || biased_exponent > MAX_BIASED_EXPONENT
        {
            return None;
        }
        Some(Self(
            ((biased_exponent as u32) << SIGNIFICAND_BITS) | significand,
        ))
    }

    /// Construct from a significand and **unbiased** exponent (`[-16, 15]`).
    #[inline]
    pub const fn encode(significand: u32, unbiased_exponent: i8) -> Option<Self> {
        let min = 0 - BIAS as i8;
        let max = MAX_BIASED_EXPONENT as i8 - BIAS as i8;
        if unbiased_exponent < min || unbiased_exponent > max {
            return None;
        }
        Self::new(significand, (unbiased_exponent as i16 + BIAS as i16) as u8)
    }

    /// Normalize a scaled significand + biased exponent into canonical
    /// bits. Truncates toward zero; returns `None` on exponent overflow.
    pub fn from_scaled(mut sig: u64, mut biased_exp: i16) -> Option<Self> {
        if sig == 0 {
            return Some(Self::ZERO);
        }
        while sig > SIGNIFICAND_MAX as u64 {
            sig /= 10;
            biased_exp += 1;
        }
        while sig < SIGNIFICAND_MIN as u64 {
            sig *= 10;
            biased_exp -= 1;
        }
        if biased_exp < 0 || biased_exp > MAX_BIASED_EXPONENT as i16 {
            return None;
        }
        Self::new(sig as u32, biased_exp as u8)
    }

    #[inline(always)]
    pub const fn is_zero(self) -> bool {
        self.0 == 0
    }

    #[inline(always)]
    pub const fn is_infinity(self) -> bool {
        self.0 == u32::MAX
    }

    /// Whether this is a valid encoding (sentinel or in-range significand).
    #[inline]
    pub const fn is_valid(self) -> bool {
        if self.0 == 0 || self.0 == u32::MAX {
            return true;
        }
        let sig = self.significand();
        sig >= SIGNIFICAND_MIN && sig <= SIGNIFICAND_MAX
    }

    #[inline(always)]
    pub const fn significand(self) -> u32 {
        self.0 & SIGNIFICAND_MASK
    }

    #[inline(always)]
    pub const fn biased_exponent(self) -> u8 {
        (self.0 >> SIGNIFICAND_BITS) as u8
    }

    #[inline(always)]
    pub const fn unbiased_exponent(self) -> i8 {
        self.biased_exponent() as i8 - BIAS as i8
    }

    /// Bid-side heap key: `u32::MAX - self`, so a min-key sort yields
    /// highest-price-first for bids.
    #[inline(always)]
    pub const fn bid_key(self) -> u32 {
        u32::MAX - self.0
    }

    /// `base * price`, rounded toward zero. `ZERO -> 0`, `INFINITY -> u128::MAX`.
    #[inline]
    pub fn quote_for_base(self, base: u64) -> u128 {
        if self.is_zero() {
            return 0;
        }
        if self.is_infinity() {
            return u128::MAX;
        }
        let sig = self.significand() as u128;
        let unb = self.unbiased_exponent() as i32 - 7;
        let mut x = (base as u128).saturating_mul(sig);
        if unb >= 0 {
            for _ in 0..unb {
                x = x.saturating_mul(10);
            }
        } else {
            for _ in 0..(-unb) {
                x /= 10;
            }
        }
        x
    }

    /// `quote / price`, rounded toward zero. `ZERO -> u128::MAX`, `INFINITY -> 0`.
    #[inline]
    pub fn base_for_quote(self, quote: u64) -> u128 {
        if self.is_zero() {
            return u128::MAX;
        }
        if self.is_infinity() {
            return 0;
        }
        let sig = self.significand() as u128;
        let unb = self.unbiased_exponent() as i32 - 7;
        let mut num = quote as u128;
        let mut den = sig;
        if unb >= 0 {
            for _ in 0..unb {
                den = den.saturating_mul(10);
            }
        } else {
            for _ in 0..(-unb) {
                num = num.saturating_mul(10);
            }
        }
        if den == 0 {
            return 0;
        }
        num / den
    }

    /// Decode to `f64`. Display/diagnostics only — never used in fill math.
    pub fn to_f64(self) -> f64 {
        if self.is_zero() {
            return 0.0;
        }
        if self.is_infinity() {
            return f64::INFINITY;
        }
        let sig = self.significand() as f64;
        let shift = self.biased_exponent() as i32 - BIAS as i32 - 7;
        sig * 10f64.powi(shift)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sentinels() {
        assert!(Price::ZERO.is_zero() && Price::ZERO.is_valid());
        assert!(Price::INFINITY.is_infinity() && Price::INFINITY.is_valid());
        assert_eq!(Price::ZERO.bid_key(), u32::MAX);
        assert_eq!(Price::INFINITY.bid_key(), 0);
    }

    #[test]
    fn fx_ratio_math() {
        // EUR/USD 1.0850
        let p = Price::encode(10_850_000, 0).unwrap();
        assert_eq!(p.quote_for_base(1_000_000), 1_085_000);
        assert_eq!(p.base_for_quote(1_085_000), 1_000_000);
        assert!((p.to_f64() - 1.085).abs() < 1e-9);
    }

    #[test]
    fn integer_order_is_price_order() {
        let a = Price::encode(99_999_999, 0).unwrap(); // 9.9999999
        let b = Price::encode(10_000_000, 1).unwrap(); // 10.0
        assert!(a < b && a.as_u32() < b.as_u32());
    }

    #[test]
    fn from_scaled_truncates() {
        let p = Price::from_scaled(123_456_789, 10).unwrap();
        assert_eq!(p.significand(), 12_345_678);
        assert_eq!(p.biased_exponent(), 11);
    }
}
